src/fluree/db/query/exec/select.cljc

```diff
@@ -0,0 +1,136 @@
+
+(defprotocol ValueSelector
+  "Format a where search solution (map of pattern matches) by extracting and
+  displaying relevant pattern matches."
+  (format-value [fmt db iri-cache compact error-ch solution]))
+
+(defrecord VariableSelector [var]
+  ValueSelector
+  (format-value
+    [_ db iri-cache compact error-ch solution]
+    (-> solution
+        (get var)
+        (display db iri-cache compact error-ch))))
+
+(defn variable-selector
+  "Returns a selector that extracts and formats a value bound to the specified
+  `variable` in where search solutions for presentation."
+  [variable]
+  (->VariableSelector variable))
+
+(defrecord AggregateSelector [fmt agg-fn]
+  ValueSelector
+  (format-value
+    [_ db iri-cache compact error-ch solution]
+    (let [agg-ch (chan 1 (map agg-fn))]
+      (-> (format-value fmt db iri-cache compact error-ch solution)
+          (async/pipe agg-ch)))))
+
+(defn aggregate-selector
+  "Returns a selector that extracts the grouped value bound to the specified
+  `variable` from a where solution, formats each item in the group, and
+  processes the formatted group with the supplied `agg-function` to generate the
+  final aggregated result for display."
+  [variable agg-function]
+  (let [var-fmt (variable-selector variable)]
+    (->AggregateSelector var-fmt agg-function)))
+
+(defrecord SubgraphSelector [var selection depth spec]
+  ValueSelector
+  (format-value
+    [_ db iri-cache compact error-ch solution]
+    (go
+      (let [sid (-> solution
+                    (get var)
+                    ::where/val)]
+        (try*
+         (let [flakes (<? (query-range/index-range db :spot = [sid]))]
+           ;; TODO: Replace these nils with fuel values when we turn fuel back on
+           (<? (json-ld-resp/flakes->res db iri-cache compact nil nil spec 0 flakes)))
+         (catch* e
+                 (log/error e "Error formatting subgraph for subject:" sid)
+                 (>! error-ch e)))))))
+
+(defn subgraph-selector
+  "Returns a selector that extracts the subject id bound to the supplied
+  `variable` within a where solution and extracts the subgraph containing
+  attributes and values associated with that subject specified by `selection`
+  from a database value."
+  [variable selection depth spec]
+  (->SubgraphSelector variable selection depth spec))
+
+(defn format-values
+  "Formats the values from the specified where search solution `solution`
+  according to the selector or collection of selectors specified by `selectors`"
+  [solution db iri-cache compact error-ch selectors]
+  (if (sequential? selectors)
+    (go-loop [selectors  selectors
+              values     []]
+      (if-let [selector (first selectors)]
+        (let [value (<! (format-value selector db iri-cache compact error-ch solution))]
+          (recur (rest selectors)
+                 (conj values value)))
+        values))
+    (format-value selectors db iri-cache compact error-ch solution)))
```

Comment:
This is the world's smallest nitpick: I find it mildly stressful that the args for `format-value` and `format-values` are the same, but `selectors` and `solution` are swapped

Reply:
It's not minor. I'll fix it.